pkg/lexer/lexer.go

```diff
@@ -0,0 +1,259 @@
+package lexer
+
+import (
+	"unicode"
+	"unicode/utf8"
+)
+
+type TokenType int
+
+// These are the main token types for shell
+const (
+	// Special tokens
+	ILLEGAL TokenType = iota
+	SEMICOLON
+	GREATER
+	LESS
+	GREATGREAT
+	LESSLESS
+	GREATAMPERSAND
+	LESSAMPERSAND
+	PIPE
+	AMPERSAND
+	IDENT
+	SINGLEQUOTE
+	DOUBLEQUOTE
+	BACKQUOTE
+	COMMENT
+	EOF
+)
+
+type Token struct {
+	Type TokenType
+	Val  string
+}
+
+// grammar
+// args -> IDENTS | empty
+// cmd and args -> IDENT args
+// pipe list -> PIPE cmd and args | cmd and args
+// io modifier -> GREAT IDENT | GREATGREAT IDENT | ...
+// io modifiers list -> io modifiers | empty
+// background optional -> ampersand | empty
+// cmd line -> pipe list io modifiers background optional NEWLINE | NEWLINE | ILLEGAL recovery.
+
+type Lexer struct {
+	input  string // input string
+	start  int    // start position of current token
+	pos    int    // current position in input
+	width  int    // width of the last rune read
+	Tokens chan Token
+}
+
+func Lex(input string) *Lexer {
+	l := &Lexer{
+		input:  input,
+		Tokens: make(chan Token),
+	}
+	go l.run()
+	return l
+}
+
+func (l *Lexer) run() {
+	for state := lexCommand; state != nil; {
+		state = state(l)
+	}
+	close(l.Tokens)
+}
+
+func (l *Lexer) emit(t TokenType) {
+	l.Tokens <- Token{t, l.input[l.start:l.pos]}
+	l.start = l.pos
+}
+
+func (l *Lexer) next() rune {
```

Comment:
Adding to the previous comment, using `l.ch` will not require this function to return anything.

Reply:
Also, you can use peek to your advantage here and and change the code to:
```go
func (l *Lexer) next() {
  l.ch, l.width = l.peek()
  l.pos += l.width
}
```